// hazelcast-client/lib/serialization/generic_record/FieldKind.js
"use strict";
/*
 * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldKind = void 0;
/**
 * FieldKind for Compact.
 * It is designed to be used with {@link GenericRecord.getFieldKind} API.
 *
 * Note that actual ids in {@link FieldType} and {@link FieldKind} are not matching.
 * {@link FieldType} is the old API for Portable only and only meant to be used with
 * {@link PortableReader.getFieldType} API.
 *
 * This API is currently in Beta and can change at any time.
 */
var FieldKind;
(function (FieldKind) {
    FieldKind[FieldKind["BOOLEAN"] = 0] = "BOOLEAN";
    FieldKind[FieldKind["ARRAY_OF_BOOLEAN"] = 1] = "ARRAY_OF_BOOLEAN";
    FieldKind[FieldKind["INT8"] = 2] = "INT8";
    FieldKind[FieldKind["ARRAY_OF_INT8"] = 3] = "ARRAY_OF_INT8";
    // char and array of char are not here because portable generic records is not supported yet
    // CHAR,
    // ARRAY_OF_CHAR,
    FieldKind[FieldKind["INT16"] = 6] = "INT16";
    FieldKind[FieldKind["ARRAY_OF_INT16"] = 7] = "ARRAY_OF_INT16";
    FieldKind[FieldKind["INT32"] = 8] = "INT32";
    FieldKind[FieldKind["ARRAY_OF_INT32"] = 9] = "ARRAY_OF_INT32";
    FieldKind[FieldKind["INT64"] = 10] = "INT64";
    FieldKind[FieldKind["ARRAY_OF_INT64"] = 11] = "ARRAY_OF_INT64";
    FieldKind[FieldKind["FLOAT32"] = 12] = "FLOAT32";
    FieldKind[FieldKind["ARRAY_OF_FLOAT32"] = 13] = "ARRAY_OF_FLOAT32";
    FieldKind[FieldKind["FLOAT64"] = 14] = "FLOAT64";
    FieldKind[FieldKind["ARRAY_OF_FLOAT64"] = 15] = "ARRAY_OF_FLOAT64";
    FieldKind[FieldKind["STRING"] = 16] = "STRING";
    FieldKind[FieldKind["ARRAY_OF_STRING"] = 17] = "ARRAY_OF_STRING";
    FieldKind[FieldKind["DECIMAL"] = 18] = "DECIMAL";
    FieldKind[FieldKind["ARRAY_OF_DECIMAL"] = 19] = "ARRAY_OF_DECIMAL";
    FieldKind[FieldKind["TIME"] = 20] = "TIME";
    FieldKind[FieldKind["ARRAY_OF_TIME"] = 21] = "ARRAY_OF_TIME";
    FieldKind[FieldKind["DATE"] = 22] = "DATE";
    FieldKind[FieldKind["ARRAY_OF_DATE"] = 23] = "ARRAY_OF_DATE";
    FieldKind[FieldKind["TIMESTAMP"] = 24] = "TIMESTAMP";
    FieldKind[FieldKind["ARRAY_OF_TIMESTAMP"] = 25] = "ARRAY_OF_TIMESTAMP";
    FieldKind[FieldKind["TIMESTAMP_WITH_TIMEZONE"] = 26] = "TIMESTAMP_WITH_TIMEZONE";
    FieldKind[FieldKind["ARRAY_OF_TIMESTAMP_WITH_TIMEZONE"] = 27] = "ARRAY_OF_TIMESTAMP_WITH_TIMEZONE";
    FieldKind[FieldKind["COMPACT"] = 28] = "COMPACT";
    FieldKind[FieldKind["ARRAY_OF_COMPACT"] = 29] = "ARRAY_OF_COMPACT";
    // portable and array of portable are not here because portable generic records is not supported yet
    // PORTABLE,
    // ARRAY_OF_PORTABLE,
    FieldKind[FieldKind["NULLABLE_BOOLEAN"] = 32] = "NULLABLE_BOOLEAN";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_BOOLEAN"] = 33] = "ARRAY_OF_NULLABLE_BOOLEAN";
    FieldKind[FieldKind["NULLABLE_INT8"] = 34] = "NULLABLE_INT8";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_INT8"] = 35] = "ARRAY_OF_NULLABLE_INT8";
    FieldKind[FieldKind["NULLABLE_INT16"] = 36] = "NULLABLE_INT16";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_INT16"] = 37] = "ARRAY_OF_NULLABLE_INT16";
    FieldKind[FieldKind["NULLABLE_INT32"] = 38] = "NULLABLE_INT32";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_INT32"] = 39] = "ARRAY_OF_NULLABLE_INT32";
    FieldKind[FieldKind["NULLABLE_INT64"] = 40] = "NULLABLE_INT64";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_INT64"] = 41] = "ARRAY_OF_NULLABLE_INT64";
    FieldKind[FieldKind["NULLABLE_FLOAT32"] = 42] = "NULLABLE_FLOAT32";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_FLOAT32"] = 43] = "ARRAY_OF_NULLABLE_FLOAT32";
    FieldKind[FieldKind["NULLABLE_FLOAT64"] = 44] = "NULLABLE_FLOAT64";
    FieldKind[FieldKind["ARRAY_OF_NULLABLE_FLOAT64"] = 45] = "ARRAY_OF_NULLABLE_FLOAT64";
    FieldKind[FieldKind["NOT_AVAILABLE"] = 46] = "NOT_AVAILABLE";
})(FieldKind = exports.FieldKind || (exports.FieldKind = {}));
